### PR TITLE
Allow narrowing `{}` originating in `unknown` using `instanceof`

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/51007

### CODE REVIEW

File: tests/cases/compiler/inKeywordAndUnknown.ts [hunk 1]
```diff
@@ -16,3 +16,9 @@ function f(x: {}, y: unknown) {
     }
     y;  // {}
 }
+
+
+// repro #51007
+function isHTMLTable(table: unknown): boolean {
+    return !!table && table instanceof Object && 'html' in table;
```
Comment (DanielRosenwasser) [inline tests/cases/compiler/inKeywordAndUnknown.ts:23]: I think the test here should actually test for an access of `table.html` in some capacity - e.g.

```ts
function tryGetHtmlPropFromTable(table: unknown) {
    if (!!table && table instanceof Object && 'html' in table && table.html instanceof Object) {
        return table.html;
    }
    return undefined;
}

Comment (Andarist) [reply tests/cases/compiler/inKeywordAndUnknown.ts:23]: hm, what would it change here? the reported error was that `in` couldn't be used at all in this situation - we could add more "assertions" here than what was originally reported but I'm not sure what this particular snippet would bring to the table here. What kind of nuance am I missing? 😅

Comment (DanielRosenwasser) [reply tests/cases/compiler/inKeywordAndUnknown.ts:23]: The original test is fine; having another test that actually tries to narrow and use the newly-tested property corresponds to something closer to real-world usage.

Comment (Andarist) [reply tests/cases/compiler/inKeywordAndUnknown.ts:23]: Ok, I've pushed out the proposed test case - please take a look if this is what you had in mind here.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -25891,6 +25891,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 if (!nonConstructorTypeInUnion) return type;
             }
 
+            if (assumeTrue && declaredType === unknownType && isEmptyAnonymousObjectType(type)) {
+                return targetType;
+            }
+
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:25897]: The problem with conditioning this on the declared type is that it doesn't work when the empty anonymous object type originates in some other type. For example:

```ts
function f<T>(x: T & {}) {
    return x instanceof Object && 'a' in x;  // Error, but shouldn't be
}
```

**Review by ahejlsberg — COMMENTED** [review]
I have what I think is the correct fix in #51631.

**Comment by Andarist** [comment]
Closing as superseded by https://github.com/microsoft/TypeScript/pull/51631

**Pull request closed without merging**